sphinx/builders/linkcheck.py

```diff
@@ -268,53 +283,99 @@ def write_linkstat(self, data: dict) -> None:
+                                                      self.rqueue, self.wqueue,
+                                                      self.rate_limits, self.builder)
+            thread.start()
+            self.workers.append(thread)
+
+    def shutdown_threads(self) -> None:
         self.wqueue.join()
-        # Shutdown threads.
         for worker in self.workers:
             self.wqueue.put((CHECK_IMMEDIATELY, None, None, None), False)
 
+    def check(self, hyperlinks: Dict[str, Hyperlink]) -> Generator[Tuple[str, str, int, str, str, int], None, None]:  # NOQA
+        self.invoke_threads()
+
+        total_links = 0
+        for hyperlink in hyperlinks.values():
+            if self.is_ignored_uri(hyperlink.uri):
+                yield CheckResult(hyperlink.uri, hyperlink.docname, hyperlink.lineno,
+                                  'ignored', '', 0)
+            else:
+                self.wqueue.put(hyperlink, False)
+                total_links += 1
+
+        done = 0
+        while done < total_links:
+            yield self.rqueue.get()
+            done += 1
+
+        self.shutdown_threads()
+
+    def is_ignored_uri(self, uri: str) -> bool:
+        return any(pat.match(uri) for pat in self.to_ignore)
+
 
 class HyperlinkAvailabilityCheckWorker(Thread):
     """A worker class for checking the availability of hyperlinks."""
 
-    def __init__(self, builder: CheckExternalLinksBuilder) -> None:
-        self.config = builder.config
-        self.env = builder.env
-        self.rate_limits = builder.rate_limits
-        self.rqueue = builder.rqueue
-        self.wqueue = builder.wqueue
+    def __init__(self, env: BuildEnvironment, config: Config, rqueue: queue.Queue,
+                 wqueue: queue.Queue, rate_limits: Dict[str, RateLimit],
+                 builder: CheckExternalLinksBuilder = None) -> None:
+        # Warning: builder argument will be removed in the sphinx-5.0.
+        # Don't use it from extensions.
+        # tag: RemovedInSphinx50Warning
+        self.config = config
+        self.env = env
+        self.rate_limits = rate_limits
+        self.rqueue = rqueue
+        self.wqueue = wqueue
 
         self.anchors_ignore = [re.compile(x)
                                for x in self.config.linkcheck_anchors_ignore]
         self.auth = [(re.compile(pattern), auth_info) for pattern, auth_info
                      in self.config.linkcheck_auth]
-        self.to_ignore = [re.compile(x) for x in self.config.linkcheck_ignore]
 
-        self._good = builder._good
-        self._broken = builder._broken
-        self._redirected = builder._redirected
+        if builder:
+            # if given, fill the result of checks as cache
+            self._good = builder._good
+            self._broken = builder._broken
+            self._redirected = builder._redirected
+        else:
+            self._good = set()
+            self._broken = {}
+            self._redirected = {}
```

Comment:
Perhaps add a comment to remove these attributes in version 5. Though it repeats the deprecated.rst document, it’s much more obvious to contributors reading the source.

Reply:
Done